internal/uvm/vnuma.go

```diff
@@ -0,0 +1,197 @@
+	var virtualSocketSet = make(map[uint32]struct{})
+	var totalVPCount uint32
+	var totalMemInMb uint64
+	var highestVNodeNumber uint32
+	var highestVSocketNumber uint32
+
+	hasWildcardPhysicalNode := n.Settings[0].PhysicalNodeNumber == WildcardPhysicalNodeNumber
+
+	for _, topology := range n.Settings {
+		if topology.VirtualNodeNumber > NumaChildNodeCountMax {
+			return fmt.Errorf("vNUMA virtual node number %d exceeds maximum allowed value %d", topology.VirtualNodeNumber, NumaChildNodeCountMax)
+		}
+		if topology.PhysicalNodeNumber != WildcardPhysicalNodeNumber && topology.PhysicalNodeNumber >= NumaTopologyNodeCountMax {
+			return fmt.Errorf("vNUMA physical node number %d exceeds maximum allowed value %d", topology.PhysicalNodeNumber, NumaTopologyNodeCountMax)
+		}
+		if hasWildcardPhysicalNode != (topology.PhysicalNodeNumber == WildcardPhysicalNodeNumber) {
+			return fmt.Errorf("vNUMA has a mix of wildcard (%d) and non-wildcard (%d) physical node numbers", WildcardPhysicalNodeNumber, topology.PhysicalNodeNumber)
+		}
+
+		if topology.CountOfMemoryBlocks == 0 {
+			return fmt.Errorf("vNUMA nodes with no memory are not allowed")
+		}
+
+		totalVPCount += topology.CountOfProcessors
+		totalMemInMb += topology.CountOfMemoryBlocks
+
+		if _, ok := virtualNodeSet[topology.VirtualNodeNumber]; ok {
+			return fmt.Errorf("vNUMA virtual node number %d is duplicated", topology.VirtualNodeNumber)
+		}
+		virtualNodeSet[topology.VirtualNodeNumber] = struct{}{}
+
+		if topology.MemoryBackingType != hcsschema.MemoryBackingType_PHYSICAL && topology.MemoryBackingType != hcsschema.MemoryBackingType_VIRTUAL {
+			return fmt.Errorf("vNUMA memory backing type %s is invalid", topology.MemoryBackingType)
+		}
+
+		if highestVNodeNumber < topology.VirtualNodeNumber {
+			highestVNodeNumber = topology.VirtualNodeNumber
+		}
+		if highestVSocketNumber < topology.VirtualSocketNumber {
+			highestVSocketNumber = topology.VirtualSocketNumber
+		}
+
+		virtualSocketSet[topology.VirtualSocketNumber] = struct{}{}
+	}
+
+	// Either both total memory and processor count should be zero or both should be non-zero
+	if (totalMemInMb == 0) != (totalVPCount == 0) {
+		return fmt.Errorf("partial resource allocation is not allowed")
+	}
+
+	// At least
+	if totalMemInMb == 0 && hasWildcardPhysicalNode {
+		return fmt.Errorf("completely empty topology is not allowed")
+	}
+
+	if len(virtualNodeSet) != int(highestVNodeNumber+1) {
+		return fmt.Errorf("holes in vNUMA node numbers are not allowed")
+	}
+
+	if len(virtualSocketSet) != int(highestVSocketNumber+1) {
+		return fmt.Errorf("holes in vNUMA socket numbers are not allowed")
+	}
+	return nil
+}
+
+// ValidateNumaForVM validates the NUMA settings for a VM with the given memory settings `memorySettings`,
+// processor count `procCount`, and total memory in MB `memInMb`.
+func ValidateNumaForVM(numa *hcsschema.Numa, vmMemoryBackingType hcsschema.MemoryBackingType, procCount uint32, memInMb uint64) error {
```

Comment:
I don't think this needs to be capitalized since it's not used outside the package

Reply:
Done.